maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,107 +137,224 @@ public fun GoogleMap(
+
+                    override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
+                        if(event == Lifecycle.Event.ON_DESTROY) {
+                            // Only destroy from onRelease. Move down to CREATED state.
+                            moveToLifecycleState(Lifecycle.State.CREATED)
+                            return
+                        }
+                        moveToLifecycleState(event.targetState)
+                    }
+
+                    @Synchronized
+                    fun moveToLifecycleState(targetState: Lifecycle.State) {
+                        while(currentLifecycleState != targetState) {
+                            if(targetState < currentLifecycleState) moveBackward()
+                            else moveForward()
+                        }
+                    }
+
+                    private fun moveBackward() {
+                        val event = Lifecycle.Event.downFrom(currentLifecycleState)
+                            ?: error("no event down from $currentLifecycleState")
+                        invokeEvent(event)
+                    }
+
+                    private fun moveForward() {
+                        val event = Lifecycle.Event.upFrom(currentLifecycleState)
+                            ?: error("no event up from $currentLifecycleState")
+                        invokeEvent(event)
+                    }
+
+                    private fun invokeEvent(event: Lifecycle.Event) {
+                        log("Invoking Lifecycle event $event")
+                        when(event) {
+                            Lifecycle.Event.ON_CREATE -> mapView.onCreate(Bundle())
+                            Lifecycle.Event.ON_START -> mapView.onStart()
+                            Lifecycle.Event.ON_RESUME -> mapView.onResume()
+                            Lifecycle.Event.ON_PAUSE -> mapView.onPause()
+                            Lifecycle.Event.ON_STOP -> mapView.onStop()
+                            Lifecycle.Event.ON_DESTROY -> mapView.onDestroy()
+                            else -> error("Unsupported lifecycle event: $event")
+                        }
+                        currentLifecycleState = event.targetState
+                    }
                 }
+
+                var lifecycleOwner by Delegates.notNull<LifecycleOwner>()
+
+                fun unregisterLifecycleObserver() {
+                    log("Unregistering lifecycle observer")
+                    lifecycleOwner.lifecycle.removeObserver(lifecycleObserver)
+                }
+
+                val attachStateListener = object : View.OnAttachStateChangeListener {
+                    override fun onViewAttachedToWindow(v: View) {
+                        log("View attached!")
+                        lifecycleOwner = v.findViewTreeLifecycleOwner()!!
+                        lifecycleOwner.lifecycle.addObserver(lifecycleObserver)
+                    }
```

Comment:
I'm seeing now that we also need a special case in the LifecycleEventObserver class to prevent moving down to INITIALIZED state (which is not valid). A changed lifecycleOwner could be INITIALIZED, after we switched from another that had reached at least CREATED.

Reply:
I'm starting to spew nonsense. Ignore that last one, please.